content/modules/python/pages/writing_functions.adoc

```diff
@@ -0,0 +1,462 @@
+----
+help(word_count)
+----
+
+[source, Python]
+----
+Help on function word_count in module __main__:
+ 
+word_count(sentence: str) -> int
+word_count is a function that accepts a sentence as an argument,
+and returns the number of words in the sentence.
+     
+Args:
+sentence (str): The sentence for which we are counting the words.
+     
+Returns:
+int: The number of words in the sentence
+----
+
+In addition, if you are coding in a tool like VSCode, for example, you may gain the ability to hover over a function and see its docstring and other information.
+
+It is good practice to write docstrings for every function you write.
+
+== annotations
+
+Another "thing" you may have noticed from our `word_count` function if you've ever used Python in the past. In the _signature_ of our function `def` `word_count(sentence: str) -> int`, we have some extra, _not_ required information in the form of _function annotations_. Specifically, you could write the `word_count` function like this, but it would function just the same:
+
+[source, Python]
+----
+def word_count(sentence):
+    """
+    word_count is a function that accepts a sentence as an argument,
+    and returns the number of words in the sentence.
+
+    Args:
+        sentence (str): The sentence for which we are counting the words.
+
+    Returns:
+        int: The number of words in the sentence
+    """
+    
+    def _strip_punctuation(sentence: str):
+        """
+        helper function to strip punctuation.
+        """
+        return sentence.translate(str.maketrans('', '', string.punctuation))
+    
+    sentence_no_punc = _strip_punctuation(sentence)
+    result = len(sentence_no_punc.split())
+    return result
+----
+
+Here, we do not specify that `sentence` should be a `str` or that the returned result should be an `int`. When we _do_ annotate functions, it is purely a way to add metadata to our function. In large projects, function annotations are recommended. Although Python does not strictly enforce type annotations, packages like `mypy` can be added to a deployment scheme to strictly enforce it.
+
+== decorators
+
+=== Examples
+
+*Write a function called get_filename_from_url that, given a url to a file, like https://image.shutterstock.com/image-vector/cute-dogs-line-art-border-260nw-1079902403.jpg returns the filename with the extension.*
+
+[source, Python]
+----
+import os
+from urllib.parse import urlparse
+
+def get_filename_from_url(url: str) -> str:
+    """
+    Given a link to a file, return the filename with extension.
+
+    Args:
+        url (str): The url of the file.
+
+    Returns:
+        str: A string with the filename, including the file extension.
+    """
+    return os.path.basename(urlparse(url).path)
+----
+
+*Write a function that, given a url to an image, and a full path to a directory, saves the image to the provided directory. By default, have the function save the images to the user's home directory in a unix-like operating system.*
```

Comment:
The function below is called `scrape_image`. It accepts a URL to an image and a full path to a directory as arguments. The function saves the image to the provided directory. If a path is not provided, (by default) the function will save the image to the user's home directory (in a UNIX-like operating system).